### PR TITLE
feat: Stacks monorepo refactor

### LINKED ISSUE #798: Refactor auth to be compatible with new authenticator and connect

### LINKED ISSUE #799: Refactor into mono repo
→ auth
        All authentication related API
        `userSession`, `signIn`
    → storage
        All Gaia related API 
        `putFile`, `getFile` etc.
    → profile
        Functions used to manipulate user profiles on BNS
    → keychain
        Stacks key generation, manipulation and addresses
        HD wallet derivation for account and wallet
    → transactions
        Transaction building API in `stacks-transactions`
    → network
        API for communicating with Stacks node and sidecar
        Including transaction broadcast, nonce, balance and fee estimation currently part of transactions lib.
    → encryption
        All existing encryption functions in blockstack.js
    → cli
        The stacks CLI, superseding the Blockstack CLI

### LINKED ISSUE #808: Refactor storage to remove circular dependencies with auth
`auth.ts` and `userSession.ts` both have dependencies on storage. Either all storage functions should be accessed through `userSession` or storage functions should be separate and take a `userSession` object as input.

### LINKED ISSUE #809: Refactor encryption and keys.ts into crypto package
Merge keys.ts with encryption functions into a `crypto` package. (A separate `keychain` package will contain logic for key derivation and etc.)

### LINKED ISSUE #810: Create network package
Create new package from https://github.com/blockstack/stacks-transactions-js/blob/master/src/network.ts. Merge with non-deprecated logic in `network.ts`.

Additionally, the Stacks node API client should live in this package.

### LINKED ISSUE #811: Create keychain package
Transfer over from https://github.com/blockstack/ux/tree/master/packages/keychain and merge with functions from https://github.com/blockstack/stacks-transactions-js/blob/master/src/keys.ts

### LINKED ISSUE #812: Create transactions package
Transfer from https://github.com/blockstack/stacks-transactions-js

### LINKED ISSUE #813: Create profiles package
Create profiles package from https://github.com/blockstack/blockstack.js/tree/master/src/profiles

### LINKED ISSUE #814: Create CLI package
Move https://github.com/blockstack/cli-blockstack/pull/40 to new CLI package.

### PR DESCRIPTION
**Pull request by yknl**
This PR splits blockstack.js into multiple NPM packages hosted from a single GitHub monorepo. See original proposal on the forum: https://forum.blockstack.org/t/stacks-mono-repo-rfc/11017

The following functionality have been removed and will be superceded by Stacks 2.0 compatible libraries/packages in this repo.
- BTC & Stacks 1.0 transaction building including BNS operations
- Stacks 1.0 wallet class
- Stacks 1.0 network class

# New packages

## Auth - @stacks/auth
A package that contains all functions relating to authentication. The API has been updated to support new implementations of authenticators and Connect. Functions depending on a legacy Blockstack browser have been dropped in favour of Connect.

## Storage - @stacks/storage
Storage functions have been separated from auth and the UserSession class. In order to use storage, you should import the `Storage` class from the package and create an instance of it using a valid `UserSession` object.

```typescript
import { UserSession } from '@stacks/auth'
import { Storage } from '@stacks/storage'

const userSession = new UserSession()
// perform authentication

const storage = new Storage(userSession)
```

## Common - @stacks/common
Utility functionality shared by packages in this monorepo have been moved to a new common package. Developers shouldn't need to include this package directly.

## Profile - @stacks/profile
API used by authenticators to manipulate user profiles on BNS have been moved to this package. Social proofs are fully deprecated.

## Encryption - @stacks/encryption
Common encryption functions used across the platform. 

## Keychain - @stacks/keychain
This package exposes the identity, key management functions. Moved from https://github.com/blockstack/ux/tree/master/packages/keychain

## Transaction - @stacks/transaction
This package contains methods used to construct all types of transactions for Stacks 2.0. Migrated from https://github.com/blockstack/stacks-transactions-js

## Network - @stacks/network
This package contains network functions for Stacks 2.0. 

## CLI - @stacks/cli
The Stacks 2.0 CLI tool. Used to create/broadcast Stacks 2.0 transactions from the command line. Moved from: https://github.com/blockstack/cli-blockstack/pull/40

# Additional Review Notes
## Build instructions
`lerna bootstrap`
`lerna run build`

## Tests
All tests have been migrated to jest for consistency. Many of the tests were updated to use jest mock and jest-fetch-mock.

### CODE REVIEW

File: migration-guide.md [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+  finished: ({ userSession }) => {
+    console.log(userSession.loadUserData());
+  },
+  appDetails: {
+    name: 'My Cool App',
+    icon: 'https://example.com/icon.png',
+  },
+};
+
+const App = () => <Connect authOptions={authOptions}>// the rest of your app's components</Connect>;
```
Comment (zone117x) [inline migration-guide.md]: Is there a clean, non-JSX alternative here?

Comment (yknl) [reply migration-guide.md]: Yes, the non-JSX alternative is using `showBlockstackConnect(authOptions)` as described below. I'll clarify.

File: .circleci/config.yml [hunk 2]
```diff
@@ -20,27 +20,27 @@ jobs:
...
+      #- run: npm run lint:formatting
 
-      - run: npm test
+      - run: npx lerna run build
 
-      - run: npm run build
+      - run: npx lerna run test
 
-      - run: npm run codecovUpload
+      #- run: npm run codecovUpload
```
Comment (zone117x) [inline .circleci/config.yml]: I think code coverage reporting for this PR is pretty important

**Comment by zone117x** [comment]
Code code reports are failing to upload. See report only showing files from old repo structure: https://codecov.io/gh/blockstack/blockstack.js/tree/ce733d62f720db6d59247453bf3bb2942bcb7446/src

The codecov upload script failure can be seen here: https://app.circleci.com/pipelines/github/blockstack/blockstack.js/125/workflows/f82b3bcb-d12c-4203-a9bd-ecf505ea9237/jobs/2087

```
==> Building file structure
==> Generating gcov reports (skip via --disable=gcov)
    Failed to run gcov command.
==> Scanning for reports
    X Failed to read file at 
==> Uploading reports
    Success!
```

It looks like coverage reports are found and uploaded for an initial set of repos, but last step fails and seems to upload an empty report? Not sure.

File: packages/auth/package.json [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    "jest-fetch-mock": "^3.0.3",
+    "ts-jest": "^26.4.1",
+    "tsdx": "^0.14.0"
+  },
+  "sideEffects": false,
+  "main": "dist/index.js",
+  "module": "dist/index.esm.js",
+  "typings": "dist/auth/src/index.d.ts",
+  "umd:main": "dist/auth.umd.production.js",
+  "unpkg": "dist/auth.cjs.production.min.js"
```
Comment (zone117x) [inline packages/auth/package.json]: Shouldn't `unpkg` point to `umd` builds?

File: packages/auth/tsconfig.json [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+      "@stacks/common": ["../../common/src"],
+      "@stacks/encryption": ["../../encryption/src"],
+      "@stacks/keychain": ["../../keychain/src"],
+      "@stacks/network": ["../../network/src"],
+      "@stacks/profile": ["../../profile/src"],
+      "@stacks/storage": ["../../storage/src"],
+      "@stacks/transactions": ["../../transactions/src"]
+    }
+  },
+  "include": ["src/**/*", "tests/**/*"]
```
Comment (zone117x) [inline packages/auth/tsconfig.json:46]: I think this tsconfig would result in `test` files being output in the `/dist` directory and being published within the npm package.

File: packages/auth/package.json (hunk 3 above)
Comment (yknl) [reply packages/auth/package.json]: It looks like publishing from lerna updates this value automatically. Not sure where this is configured, we might have to update manually.

File: packages/auth/tsconfig.json (hunk 4 above)
Comment (yknl) [reply packages/auth/tsconfig.json:46]: That doesn't seem to be the case, I wasn't able to find test files in the published packages.

**Comment by zone117x** [comment]
@yknl are we still generating reference docs with typedoc (or any other tool)?

**Comment by yknl** [comment]
> @yknl are we still generating reference docs with typedoc (or any other tool)?

Yes, still using typedoc. See the latest pushed commit.

**Review by zone117x — APPROVED** [review]
Haven't had a chance to test out the new dev setup myself yet, but project structure and code looks good to me. Great work @yknl!

**Pull request merged**